extensions/amp-a4a/0.1/friendly-frame-renderer.js

```diff
@@ -0,0 +1,106 @@
+ */
+
+import {A4AVariableSource} from '../../amp-a4a/0.1/a4a-variable-source';
+import {Renderer} from './amp-ad-type-defs';
+import {createElementWithAttributes} from '../../../src/dom';
+import {dev} from '../../../src/log';
+import {dict} from '../../../src/utils/object';
+import {getDefaultBootstrapBaseUrl} from '../../../src/3p-frame';
+import {
+  installFriendlyIframeEmbed,
+  setFriendlyIframeEmbedVisible,
+} from '../../../src/friendly-iframe-embed';
+import {installUrlReplacementsForEmbed} from '../../../src/service/url-replacements-impl';
+import {setStyle} from '../../../src/style';
+
+/**
+ * Render a validated AMP creative directly in the parent page.
+ */
+export class FriendlyFrameRenderer extends Renderer {
+
+  constructor() {
+    super();
+
+    /**
+     * @type {?Element}
+     * @private
+     */
+    this.iframe_ = null;
+  }
+
+  /** @override */
+  render(context, element, creativeData) {
+    const size = context.size;
+    const adUrl = context.requestUrl;
+    const creativeMetaData = creativeData.creativeMetaData;
+
+    dev().assert(size, 'missing creative size');
+    dev().assert(adUrl, 'missing ad request url');
+
+    // Create and setup friendly iframe.
+    this.iframe_ = /** @type {!HTMLIFrameElement} */(
+      createElementWithAttributes(
+          /** @type {!Document} */(element.ownerDocument),
+          'iframe',
+          dict({
+            // NOTE: It is possible for either width or height to be 'auto',
+            // a non-numeric value.
+            'height': size.height,
+            'width': size.width,
+            'frameborder': '0',
+            'allowfullscreen': '',
+            'allowtransparency': '',
+            'scrolling': 'no',
+          })));
+    context.applyFillContent(this.iframe_);
+
+    const fontsArray = [];
+    if (creativeMetaData.customStylesheets) {
+      creativeMetaData.customStylesheets.forEach(s => {
+        const href = s['href'];
+        if (href) {
+          fontsArray.push(href);
+        }
+      });
+    }
+
+    return installFriendlyIframeEmbed(
+        this.iframe_, element, {
+          host: element,
+          url: /** @type {string} */ (adUrl),
+          html: creativeMetaData.minifiedCreative,
+          extensionIds: creativeMetaData.customElementExtensions || [],
+          fonts: fontsArray,
+        }, embedWin => {
+          installUrlReplacementsForEmbed(context.ampDoc, embedWin,
+              new A4AVariableSource(context.ampDoc, embedWin));
+        })
+        .then(friendlyIframeEmbed => {
+          setFriendlyIframeEmbedVisible(
+              friendlyIframeEmbed, context.isInViewport());
+          // Ensure visibility hidden has been removed (set by boilerplate).
+          const frameDoc = friendlyIframeEmbed.iframe.contentDocument ||
+              friendlyIframeEmbed.win.document;
+          setStyle(frameDoc.body, 'visibility', 'visible');
+        });
+  }
+
+  /** @return {!Element} */
+  getIframe() {
```

Comment:
>The reason I need this at all is because I expect the template validator will override this class, and it needs access to the iframe.

Sorry I didn't quite get it. What's the reason for template validator to override friendly-iframe renderer? 
And let's only add this getter method when necessary. Thanks.

Reply:
Sorry, I meant template _renderer_ will subclass this class, since it's essentially rendering into a friendly frame with some additional logic (template macro rendering and such).

Reply:
I see. Cool